snuba/datasets/errors_replacer.py

```diff
@@ -176,37 +245,42 @@ def process_message(self, message: ReplacementMessage) -> Optional[Replacement]:
                 event, self.__required_columns, self.__state_name
             )
         elif type_ == "replace_group":
-            processed = process_replace_group(
-                event, self.__all_columns, self.__state_name
-            )
+            processed = ReplaceGroupReplacement.parse_message(event, context)
         elif type_ == "exclude_groups":
-            processed = process_exclude_groups(event)
+            processed = ExcludeGroupsReplacement.parse_message(event, context)
         else:
             raise InvalidMessageType("Invalid message type: {}".format(type_))
 
         return processed
 
-    def pre_replacement(self, replacement: Replacement, matching_records: int) -> bool:
-        # query_time_flags == (type, project_id, [...data...])
-        flag_type, project_id = replacement.query_time_flags[:2]
+    def pre_replacement(  # type: ignore[override]
+        self, replacement: Replacement, matching_records: int
```

Comment:
liskov substitution principle!

what I actually want here is an associated type, like in Rust:

```
impl ReplacementProcessor for ErrorsReplacer {
    type Replacement = ErrorReplacement;

    fn pre_replacement(&self, replacement: Self::Replacement) -> ...
    
}
```

alternatively I can probably make it work by putting a generic on ReplacementProcessor like I did with Replacement in prev commit

Reply:
> alternatively I can probably make it work by putting a generic on ReplacementProcessor like I did with Replacement in prev commit

I think the generic idea would work and be clean. `ReplacementProcessor` only defines the interface based on the `Repalcement` class, so you can easily make it a parameter. You can make it bounded to the Replacement class but that would not be required.